### PR TITLE
[CryptoAuthLib provider] Implementation of export key operation

### PR DESCRIPTION
**Pull request by TomaszPawelecGL**
- implementation delivered on behalf of Robert Drazkowski
- CI does not pass. It seems not to be related to this delivery.  Problem visible in the logs for picky-asn1-x509.

### CODE REVIEW

**Review by ionut-arm — COMMENTED** [review]
Thanks for the change! Left a few comments, but many might be because I don't understand the CAL interface properly.

**Comment by ionut-arm** [comment]
Btw, don't worry about the CI failures, I've pushed a fix in #443 that you can rebase on once it's merged.

File: e2e_tests/tests/per_provider/normal_tests/export_key.rs
Comment (TomaszPawelecGL) [reply e2e_tests/tests/per_provider/normal_tests/export_key.rs:388]: This test is intentionally dedicated to CAL only because the CAL hardware does not allow to read private keys from slots.

File: e2e_tests/tests/per_provider/normal_tests/export_key.rs
Comment (TomaszPawelecGL) [reply e2e_tests/tests/per_provider/normal_tests/export_key.rs:425]: Across all the tests the results returned from export_key() function are assigned to variables called ‘_exported_key’. Such variables can store either a key or an error.  In this particular test case the variable stores an error and I would like to keep its name as it is to have consistency between all the tests. 

However, I agree with your comment for verification of the response code and it’s fixed.

File: config.toml
Comment (TomaszPawelecGL) [reply config.toml]: fixed

File: src/providers/cryptoauthlib/access_keys.rs
Comment (TomaszPawelecGL) [reply src/providers/cryptoauthlib/access_keys.rs:18]: There is intentional separation of concerns between CAL-provider (Parsec) and the CAL-crate. The responsibility of CAL-crate is reduced to directly handling of the crypto chip (communication, configuration, etc.). All the rest is considered as CAL-provider’s responsibility including files handlings.

File: src/providers/cryptoauthlib/access_keys.rs
Comment (TomaszPawelecGL) [reply src/providers/cryptoauthlib/access_keys.rs]: removed.

File: src/providers/cryptoauthlib/access_keys.rs
Comment (TomaszPawelecGL) [reply src/providers/cryptoauthlib/access_keys.rs:58]: It does make sense to sense to continue if there is an error here because there might be several access keys and having problem with one of them should not prevent from using the rest which are fine.

**Review by ionut-arm — APPROVED** [review]

File: src/providers/cryptoauthlib/access_keys.rs
Comment (TomaszPawelecGL) [reply src/providers/cryptoauthlib/access_keys.rs:51]: log added.

File: src/providers/cryptoauthlib/access_keys.rs
Comment (hug-dev) [inline src/providers/cryptoauthlib/access_keys.rs:22]: Just for my understanding (I don't know much about the hardware 😅), what are access keys? How are they different from normal keys and why do they need to be put in the configuration file?

**Review by hug-dev — COMMENTED** [review]
Thanks for adding this support! Globally looks good to me, I just have a few questions.

File: e2e_tests/tests/per_provider/normal_tests/export_key.rs
Comment (TomaszPawelecGL) [reply e2e_tests/tests/per_provider/normal_tests/export_key.rs]: done

File: e2e_tests/tests/per_provider/normal_tests/export_public_key.rs
Comment (TomaszPawelecGL) [reply e2e_tests/tests/per_provider/normal_tests/export_public_key.rs]: I agree that having the same name for those tests is misguiding and that’s is why I’ve modified their name (they are different).

File: src/providers/cryptoauthlib/access_keys.rs
Comment (TomaszPawelecGL) [reply src/providers/cryptoauthlib/access_keys.rs:22]: Those keys are used for encrypting communication between CAL and a host processor so they provide another level of security. 

To understand relations between those access keys and regular keys stored in CAL you could imagine a use case of transferring (importing) a regular key to CAL when the content of that key is encrypted by an access key on the host and that encrypted content is next transferred from the host to the CAL which decrypts that encrypted key at the end.

**Comment by TomaszPawelecGL** [comment]
I cannot run CI tests on my local host after commits from 16 Jun (cf146e6, b940fa0). Once I enter the following command (i.e. for cryptouthlib):
docker run -v $(pwd):/tmp/parsec -w /tmp/parsec ghcr.io/parallaxsecond/parsec-service-test-all /tmp/parsec/ci.sh --no-stress-test cryptoauthlib

I get errors such as:

_---- per_provider::normal_tests::aead::aead_encrypt_ccm_decrypt stdout ----
thread 'per_provider::normal_tests::aead::aead_encrypt_ccm_decrypt' panicked at 'called `Result::unwrap()` on an `Err` value: Client(InvalidSocketAddress)', src/lib.rs:57:72
stack backtrace:
   0: rust_begin_unwind
             at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/std/src/panicking.rs:493:5
   1: core::panicking::panic_fmt
             at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/core/src/panicking.rs:92:14
   2: core::option::expect_none_failed
             at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/core/src/option.rs:1329:5
   3: core::result::Result<T,E>::unwrap
             at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/core/src/result.rs:1037:23
   4: e2e_tests::TestClient::new
             at ./src/lib.rs:57:27
   5: mod::per_provider::normal_tests::aead::aead_encrypt_ccm_decrypt
             at ./tests/per_provider/normal_tests/aead.rs:142:22
   6: mod::per_provider::normal_tests::aead::aead_encrypt_ccm_decrypt::{{closure}}
             at ./tests/per_provider/normal_tests/aead.rs:140:1
   7: core::ops::function::FnOnce::call_once
             at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/core/src/ops/function.rs:227:5
   8: core::ops::function::FnOnce::call_once
             at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/core/src/ops/function.rs:227:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace._

I examined those commits and I found that this entry “export PARSEC_SERVICE_ENDPOINT="unix:/tmp/parsec.sock"” was removed from ci.sh and  generate-keys.sh.

Once I restored that entry in those files the CI passed.

Could you please, tell me how I can now locally execute CI tests? I still would like to have a way to check if my new code does not break anything.

**Comment by ionut-arm** [comment]
That's interesting, we did move the environment variable out of the `ci.sh` file and into the Docker image directly. It seems that maybe your local version of the Docker images haven't been updated? CI seems to work and it uses the same calls

**Comment by TomaszPawelecGL** [comment]
Thank you for the hint - it solved my problems! :-)

**Comment by TomaszPawelecGL** [comment]
Is there anything more that I should do before getting second approval and be able to merge this functionality?

I’m taking vacation soon and I would like to close this pull request before my leave.

**Comment by ionut-arm** [comment]
Hi!

> I’m taking vacation soon and I would like to close this pull request before my leave.

How soon is that? @hug-dev is back from his vacation tomorrow, so he'll be able to approve then,  but if you're leaving before that I could merge before.

**Comment by TomaszPawelecGL** [comment]
Hi, 

I’m taking leave from 12.07 and I would like to close all my pending issues. Given this pull request is a bit old (29 days) and I’m new in the project I was afraid that I won’t be ale to get it finished before my vacation.

If @hug-dev is back tomorrow and there is nothing more from me to do then I see no problem in waiting few days more. 

Thank you for your comment! :-)

**Comment by ionut-arm** [comment]
> Given this pull request is a bit old (29 days) and I’m new in the project I was afraid that I won’t be ale to get it finished before my vacation.

Yes, apologies for that, it's usually much quicker. I think we'll get it merged in tomorrow!

File: e2e_tests/tests/per_provider/normal_tests/export_public_key.rs
Comment (hug-dev) [reply e2e_tests/tests/per_provider/normal_tests/export_public_key.rs]: OK, thanks!

File: src/providers/cryptoauthlib/access_keys.rs
Comment (hug-dev) [reply src/providers/cryptoauthlib/access_keys.rs:22]: I see that makes sense, thanks!

**Review by hug-dev — APPROVED** [review]
Thanks! Sorry for this taking so long. Enjoy your vacation 👏

**Pull request merged**
